### PR TITLE
Add mediator and simple client implementation

### PR DESCRIPTION
**Pull request by nain-F49FF806**
Client agent has majorly common code with mediator agent.
It's in essence one hybrid, with code organized separately. 
This is in anticipation of splitting client into separate binary.  

Then again, Aries agents are meant to be horizontal (p2p). So perhaps a multifunctional agent isn't all that odd.

### CODE REVIEW

**Comment by Patrik-Stas** [comment]
Hi @nain-F49FF806 note that all of the commits must be DCO signed

**Comment by nain-F49FF806** [comment]
@Patrik-Stas Yes. This PR is currently WIP. (Specifically the unsigned commit)

I was using the (absence of) DCO as an automatic block. :sweat_smile:

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
review is not complete yet (far from it), but just submitting my intermediate comments

**Comment by nain-F49FF806** [comment]
Mediator now supports mediator-coordination protocol. Also included e2e tests of functionality. 
Todo(later/soon): Use "did:key" formatting/encoding as per [this recommendation](https://github.com/hyperledger/aries-rfcs/blob/main/features/0211-route-coordination/README.md#encoding-of-keys) of [Aries RFC0360](https://github.com/hyperledger/aries-rfcs/blob/main/features/0360-use-did-key/README.md)

File: mediator/src/routes/mod.rs
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply mediator/src/routes/mod.rs:1]: You can generally find a many variations of diagrams about this if you look for "onion architecture" - but don't overthinking it now, looking at all those diagrams. Starting point, well actionable, is to:
- create service layer, which will be used by the HTTP controller layer
- the service layer should also own core components such as Wallet (and even aries-vcx in general) or your own persistence layer. HTTP api should do necessary request processing and delegate the business logic to respective service object

File: mediator/src/routes/mod.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                        .pack_didcomm(&aries_response, &our_signing_key, &their_diddoc)
+                        .await?
+                }
+                GeneralAriesMessage::XumPickup(pickup_message) => {
+                    handle_pickup_protocol(agent, pickup_message).await?;
+                    todo!();
+                }
+            }
+        };
+    let EncryptionEnvelope(packed_message_bytes) = packed_response;
```
Comment (Patrik-Stas) [inline mediator/src/routes/mod.rs:160]: elaborating on my comment for this entire file, this entire section for example should be pulled into service layer

File: mediator/src/routes/mod.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                            width: 50%;
+                            height: 50%;
+                        }}
+                    </style>
+                    {oob_qr_svg} <br>
+                    <pre>{oob_string}</pre>"
+            ))
+            .into_response()
+        }
+    }
```
Comment (Patrik-Stas) [inline mediator/src/routes/mod.rs:191]: this for example should stay in this layer - it's only API interface specific thing that you want to provide rendered QR codes. 
But `oob_invite_json` should be in service layer

File: mediator/src/routes/mod.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+pub async fn readme() -> Html<String> {
+    Html("<p>Please refer to the API section of <a>readme</a> for usage. Thanks. </p>".into())
+}
+
+pub async fn build_router<T: BaseWallet + 'static, P: MediatorPersistence>(
+    agent: Agent<T, P>,
+) -> Router {
+    Router::default()
+        .route("/", get(readme))
```
Comment (Patrik-Stas) [inline mediator/src/routes/mod.rs:209]: You can make `GET /` a simple service healthcheck endpoint, I'd suggest just return `200 OK` with empty body

File: mediator/src/routes/mod.rs [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        .route("/register", get(oob_invite_qr))
+        .route("/register.json", get(oob_invite_json))
+        .route("/aries", get(handle_aries).post(handle_aries))
```
Comment (Patrik-Stas) [inline mediator/src/routes/mod.rs:212]: Let's do `/didcomm`

File: mediator/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+anyhow = "1.0.75"
+aries-vcx = { path = "../aries_vcx" }
+aries_vcx_core = { path = "../aries_vcx_core", features = ["vdrtools_wallet"] }
+async-trait = "0.1.73"
+axum = "0.6"
+axum-macros = "0.3.8"
+cursive = { version = "0.20.0", features = ["crossterm-backend"] }
```
Comment (Patrik-Stas) [inline mediator/Cargo.toml:15]: This is fairly heavy required dependency which is not needed for mediator itself
```
> cargo tree

├── cursive v0.20.0
│   ├── ahash v0.8.3 (*)
│   ├── cfg-if v1.0.0
│   ├── crossbeam-channel v0.5.8
│   │   ├── cfg-if v1.0.0
│   │   └── crossbeam-utils v0.8.16 (*)
│   ├── crossterm v0.25.0
│   │   ├── bitflags v1.3.2
│   │   ├── libc v0.2.149
│   │   ├── mio v0.8.8 (*)
│   │   ├── parking_lot v0.12.1 (*)
│   │   ├── signal-hook v0.3.17
│   │   │   ├── libc v0.2.149
│   │   │   └── signal-hook-registry v1.4.1
│   │   │       └── libc v0.2.149
│   │   └── signal-hook-mio v0.2.3
│   │       ├── libc v0.2.149
│   │       ├── mio v0.8.8 (*)
│   │       └── signal-hook v0.3.17 (*)
│   ├── cursive_core v0.3.7
│   │   ├── ahash v0.8.3 (*)
│   │   ├── crossbeam-channel v0.5.8 (*)
│   │   ├── enum-map v2.6.3
│   │   │   └── enum-map-derive v0.14.0 (proc-macro)
│   │   │       ├── proc-macro2 v1.0.69 (*)
│   │   │       ├── quote v1.0.33 (*)
│   │   │       └── syn v2.0.38 (*)
│   │   ├── enumset v1.1.3
│   │   │   └── enumset_derive v0.8.1 (proc-macro)
│   │   │       ├── darling v0.20.3 (*)
│   │   │       ├── proc-macro2 v1.0.69 (*)
│   │   │       ├── quote v1.0.33 (*)
│   │   │       └── syn v2.0.38 (*)
│   │   ├── lazy_static v1.4.0 (*)
│   │   ├── log v0.4.20 (*)
│   │   ├── num v0.4.1
│   │   │   ├── num-complex v0.4.4
│   │   │   │   └── num-traits v0.2.17 (*)
│   │   │   ├── num-integer v0.1.45 (*)
│   │   │   ├── num-iter v0.1.43
│   │   │   │   ├── num-integer v0.1.45 (*)
│   │   │   │   └── num-traits v0.2.17 (*)
│   │   │   │   [build-dependencies]
│   │   │   │   └── autocfg v1.1.0
│   │   │   ├── num-rational v0.4.1
│   │   │   │   ├── num-integer v0.1.45 (*)
│   │   │   │   └── num-traits v0.2.17 (*)
│   │   │   │   [build-dependencies]
│   │   │   │   └── autocfg v1.1.0
│   │   │   └── num-traits v0.2.17 (*)
│   │   ├── owning_ref v0.4.1
│   │   │   └── stable_deref_trait v1.2.0
│   │   ├── time v0.3.20 (*)
│   │   ├── unicode-segmentation v1.10.1
│   │   ├── unicode-width v0.1.11
│   │   └── xi-unicode v0.3.0
│   ├── lazy_static v1.4.0 (*)
│   ├── libc v0.2.149
│   ├── log v0.4.20 (*)
│   ├── maplit v1.0.2
│   ├── ncurses v5.101.0
│   │   └── libc v0.2.149
│   │   [build-dependencies]
│   │   ├── cc v1.0.83 (*)
│   │   └── pkg-config v0.3.27
│   ├── signal-hook v0.3.17 (*)
│   ├── term_size v0.3.2
│   │   └── libc v0.2.149
│   ├── unicode-segmentation v1.10.1
│   └── unicode-width v0.1.11
```

Let's make the `tui` separate project which will not be part of `mediator` project per se. If you want, we can leave this for separate PR with additional refactoring.

File: mediator/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+serde = "1.0.188"
+serde_json = "1.0.106"
+sqlx = "0.7"
+# sqlx = { version = "0.5.8", git = "https://github.com/jovfer/sqlx", branch = "feature/json_no_preserve_order_v5" }
+thiserror = "1.0.49"
+tokio = { version = "1", features = ["rt-multi-thread", "macros"] }
+tower-http = { version = "0.4.4", features = ["catch-panic"] }
+url = "2.4.1"
+uuid = "1.4.1"
+xum-test-server = { git = "https://github.com/nain-F49FF806/axum-test-server.git", branch = "aries-vcx-mediator-integration", version = "*" }
```
Comment (Patrik-Stas) [inline mediator/Cargo.toml:33]: Perhaps I've already wrote this elsewhere before - if yes apologies for duplication - but let's make sure to inline the necessary code from here so we get rid of this dependency.

File: .github/workflows/mediator.pr.yml [hunk 7]
```diff
@@ -0,0 +1,56 @@
+name: CI mediator
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - "**"
+    paths:
```
Comment (Patrik-Stas) [inline .github/workflows/mediator.pr.yml:10]: didn't know about this, pretty cool 👍 answers my just-about-to-ask question why is this a separate file

File: mediator/src/agent/client.rs
Comment (Patrik-Stas) [inline mediator/src/agent/client.rs:1]: This is a long file ;-) I'd suggest to extract shared sections of code to some "workflows" file, sequence of code which tend to be reused commonly in test. Then the test itself should be just few lines of code, reusing these "workflows". It should be also easy to look at 2 tests code and see how they differ (right now it's difficult, because each test is very long, so it's hard to tell (other than relying on correctness of the test name))

Additional note:
When tests are fast, it's nice to have granular tests for each single scenario like you do, even when there are overlaps - the benefit of doing so is accurate reporting on failures, as you the test case itself will immediately indicates what's wrong. 
On the other hand, when tests are slow and heavy, it's worthing considering have more condensed tests, combining more scenarios

So I can imagine having 1 test which list keys, adds 2 keys, list keys, remove 1 key, list keys and I think it would be reasonable test. At the same time, if tests are fast, I don't see much downside to make them granular as you did (separate test for adding, separate with removal, etc.)

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
left additional review

File: mediator/Cargo.toml (hunk 5 above)
Comment (nain-F49FF806) [reply mediator/Cargo.toml]: I have moved this (`tui`) behind a feature flag. It should not pull in the heavy dependency by default now. 
We can still do the separation from mediator code at a later point in time.

File: mediator/Cargo.toml
[the remark this answers is not in the document]
Comment (nain-F49FF806) [reply mediator/Cargo.toml:1]: Done

File: mediator/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+messages = { path = "../messages" }
+reqwest = { version = "0.11.20", features = ["json"] }
+serde = "1.0.188"
+serde_json = "1.0.106"
+thiserror = "1.0.49"
+tokio = { version = "1", features = ["rt-multi-thread", "macros"] }
+tower-http = { version = "0.4.4", features = ["catch-panic"] }
+url = "2.4.1"
+uuid = "1.4.1"
+xum-test-server = { git = "https://github.com/nain-F49FF806/axum-test-server.git", branch = "aries-vcx-mediator-integration" }
```
[the remark this answers is not in the document]
Comment (nain-F49FF806) [reply mediator/Cargo.toml]: Done. 
More can be done. I am looking into it.

File: mediator/Cargo.toml (hunk 6 above)
Comment (nain-F49FF806) [reply mediator/Cargo.toml]: Have inlined the code inside aries-vcx repository.
Currently, it still acts as a local dependency.
As I implement Forward message support, I shall try to integrate it deeper.

File: mediator/src/agent/client.rs
Comment (nain-F49FF806) [reply mediator/src/agent/client.rs:1]: Hi, I have updated the test code. They should be more readable now. 

Thanks for the note on heavy workflow tests versus smaller modular tests. 
Since I am implementing protocol support, I wanted modular tests, so we can test each individual request/response.

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    branches: [ "main", "ci-integration-test", "ci-cd", "dev", aries-vcx-mediator-integration]
+  pull_request:
+    branches: [ "main", "ci-integration-test", "ci-cd", "dev", aries-vcx-mediator-integration]
+
+env:
+  CARGO_TERM_COLOR: always
+
+jobs:
+  check:
+    name: Check for obvious errors
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:14]: There's workspace wide job `workspace_clippy`  which runs clippy on all projects in workspace. Running `cargo check` does subset of what clippy does. So this is somewhat superfluous and I think can be dropped in favor of `workspace_clippy`

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+          ~/.cargo/registry/index/
+          ~/.cargo/registry/cache/
+          ~/.cargo/git/db/
+          target/
+        key: ${{ runner.os }}-cargo-check-cache-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-
+    - name: cargo check
+      run: cargo check --verbose
+  build:
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:33]: formatting: we have convention to leave blank line between jobs, see our main GA file

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        path: |
+          ~/.cargo/bin/
+          ~/.cargo/registry/index/
+          ~/.cargo/registry/cache/
+          ~/.cargo/git/db/
+          target/
+        key: ${{ runner.os }}-cargo-build-cache-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-    
+    - name: Build
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:53]: It would be more valuable to build docker images for mediator, then publishing the binary itself. 

But hey, why not. Though note the usefulness of building this binary is limited, as you are only building for a single architecture.

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        path: |
+          ~/.cargo/bin/
+          ~/.cargo/registry/index/
+          ~/.cargo/registry/cache/
+          ~/.cargo/git/db/
+          target/
+        key: ${{ runner.os }}-cargo-lint-cache-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-        
+    - name: Run linters
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:89]: as I mentioned, we run clippy on workspace level already

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        # Available Options:
+        #    warn: Output a warning but do not fail the action
+        #    error: Fail the action with an error message
+        #    ignore: Do not output any warnings or errors, the action does not fail
+        if-no-files-found: warn # optional, default is warn
+        # Duration after which artifact will expire in days. 0 means using default retention.
+        # Minimum 1 day. Maximum 90 days unless changed from the repository settings page.
+        retention-days: 1 # optional
+  lint:
+    if: github.event_name == 'pull_request'
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:72]: This would mean we don't run this on merge, even merge to main. And we definitely want to run everything in main branch ;-)

File: agents/rust/mediator/mediation/.github/workflows/rust-basic.yml [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        key: ${{ runner.os }}-cargo-test-cache-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-
+    - name: Install prerequisites (sqlx)
+      run: cargo install sqlx-cli || true
+    - name: Setup database
+      run: DATABASE_URL=${MYSQL_URL} sqlx migrate run
+    - name: E2E test endpoints
+      run: |
+        cargo run &
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/.github/workflows/rust-basic.yml:128]: just a general note here ... this is fine, but common approach we also use is to 
1. build docker image in one job
2. in another job, load up the built image and run integration/e2e tests against that.

File: agents/rust/mediator/mediation/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+axum = "0.6.18"
+dotenvy = "0.15.7"
+env_logger = "0.10.0"
+futures = "0.3.28"
+log = "0.4.19"
+serde = { version = "1.0.164", features = ["derive"] }
+serde_json = "1.0.104"
+serde_with = { version = "3.1.0", features = ["base64"] }
+sqlx = { version = "0.7", features = ["runtime-tokio"], optional = true }
+# sqlx = { version = "0.5.8", git = "https://github.com/jovfer/sqlx", branch = "feature/json_no_preserve_order_v5", features = [ "sqlite", "mysql", "json_no_preserve_order", "runtime-tokio-rustls"], optional = true }
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/Cargo.toml:28]: great to see you managed to upgrade! Can we delete this comment now? (and other commented lines in this file)

File: agents/rust/mediator/mediation/README.md [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+Original readme is available below.
+
+> # axum-test-server
+>
+> A axum server app to test and learn Rust and Axum features.
+>
+> The commits are made in such a way as to demonstrate and try one feature or workflow (refactoring) at a time.  
+> So the git log could be educative.  
+>
+> ## Some Interesting commits
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/README.md:28]: Feel free to include this somewhere along project plan notes, but probably doesn't belong here.

File: agents/rust/mediator/mediation/src/router.rs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        .route(
+            "/",
+            get(hello_world::handle_get).post(hello_world::handle_get),
+        )
+        .route(
+            "/json",
+            get(json::echo_message_json).post(respond_message_json),
+        )
+        .route("/forward", post(handle_forward))
+        .route("/pickup", post(handle_pickup))
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/src/router.rs:33]: these can be either:
- a single `didcomm` endpoint
- or 2 endpoints as you have it, but then I'd suggest `/didcomm` for 3rd party fw messages and `/didcomm-client` for didcomm messages from clients (eg connections, mediation coordition, pickup)

2 endpoints can be really good idea from oversight perspective, in logs it would be easier to distinguish between incoming 3rd party messages, client messages and plot stats and graphs out of that data

File: agents/rust/mediator/mediation/src/routes/json.rs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+pub async fn echo_message_json(Json(body): Json<MessageJson>) -> Json<MessageJson> {
+    Json(body)
+}
+
+pub async fn respond_message_json(Json(body): Json<MessageJson>) -> Json<MessageResponseJson> {
+    Json(MessageResponseJson {
+        message: (body.message),
+        response: ("I am groot".to_owned()),
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/src/routes/json.rs:25]: 😆 👍

File: agents/rust/mediator/mediation/src/server.rs [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+use log::info;
+
+use crate::router::create_router;
+
+pub async fn run_server() {
+    // app definition, and routings
+    let app = create_router().await;
+    info!("Starting server");
+    // add server task to main loop
+    axum::Server::bind(&"127.0.0.1:7999".parse().unwrap())
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/src/server.rs:12]: Let's the IP and the port configurable. You can keep the current values as defaults

File: agents/rust/mediator/mediation/src/storage/database/mod.rs [hunk 20]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2023 Naian G.
+// SPDX-License-Identifier: Apache-2.0
+
+// unimplemented!()
```
Comment (Patrik-Stas) [inline agents/rust/mediator/mediation/src/storage/database/mod.rs:4]: if we going with mysql only, I guess we can remove the commented code?

**Comment by Patrik-Stas** [comment]
@nain-F49FF806 I've left another round of comments. I think we could get initial version merged after these, and we can keep bigger improvements in a todo list externally

**Comment by Patrik-Stas** [comment]
For now, I am merging this as is~ The outstanding comments shall be addressed in subsequent PRs
